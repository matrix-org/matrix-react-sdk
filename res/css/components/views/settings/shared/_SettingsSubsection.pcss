/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_SettingsSubsection {
    width: 100%;
    box-sizing: border-box;
}

.mx_SettingsSubsection_description {
    margin-top: $spacing-8;
}

.mx_SettingsSubsection_text {
    width: 100%;
    box-sizing: inherit;
    color: $secondary-content;
}

.mx_SettingsSubsection_content {
    width: 100%;
    display: grid;
    grid-gap: $spacing-8;
    // setting minwidth 0 makes columns definitely sized
    // fixing horizontal overflow
    grid-template-columns: minmax(0, 1fr);
    justify-items: flex-start;
    margin-top: $spacing-24;

    summary {
        color: $accent;
    }
    details[open] {
        summary {
            margin-bottom: $spacing-8;
        }
    }

    &.mx_SettingsSubsection_contentStretch {
        justify-items: stretch;
    }

    &.mx_SettingsSubsection_noHeading {
        margin-top: 0;
    }
}
