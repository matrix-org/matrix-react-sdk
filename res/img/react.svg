<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
