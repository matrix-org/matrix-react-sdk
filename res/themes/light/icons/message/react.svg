<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>