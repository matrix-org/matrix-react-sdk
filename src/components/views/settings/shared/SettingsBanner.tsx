/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren, ReactNode } from "react";

import AccessibleButton from "../../elements/AccessibleButton";

interface Props {
    icon?: ReactNode;
    action?: ReactNode;
    onAction?: () => void;
}

export function SettingsBanner({ children, icon, action, onAction }: PropsWithChildren<Props>): JSX.Element {
    return (
        <div className="mx_SettingsBanner">
            {icon}
            <div className="mx_SettingsBanner_content">{children}</div>
            {action && (
                <AccessibleButton kind="primary_outline" onClick={onAction ?? null}>
                    {action}
                </AccessibleButton>
            )}
        </div>
    );
}
