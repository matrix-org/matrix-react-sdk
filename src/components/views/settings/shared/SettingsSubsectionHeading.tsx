/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { HTMLAttributes } from "react";

import Heading from "../../typography/Heading";

export interface SettingsSubsectionHeadingProps extends HTMLAttributes<HTMLDivElement> {
    heading: string;
    legacy?: boolean;
    children?: React.ReactNode;
}

export const SettingsSubsectionHeading: React.FC<SettingsSubsectionHeadingProps> = ({
    heading,
    legacy = true,
    children,
    ...rest
}) => {
    const size = legacy ? "4" : "3";

    return (
        <div {...rest} className="mx_SettingsSubsectionHeading">
            <Heading className="mx_SettingsSubsectionHeading_heading" size={size} as="h3">
                {heading}
            </Heading>
            {children}
        </div>
    );
};
