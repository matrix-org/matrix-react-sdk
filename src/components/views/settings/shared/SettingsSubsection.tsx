/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classNames from "classnames";
import React, { HTMLAttributes } from "react";

import { SettingsSubsectionHeading } from "./SettingsSubsectionHeading";

export interface SettingsSubsectionProps extends HTMLAttributes<HTMLDivElement> {
    heading?: string | React.ReactNode;
    description?: string | React.ReactNode;
    children?: React.ReactNode;
    // when true content will be justify-items: stretch, which will make items within the section stretch to full width.
    stretchContent?: boolean;
}

export const SettingsSubsectionText: React.FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...rest }) => (
    <div {...rest} className="mx_SettingsSubsection_text">
        {children}
    </div>
);

export const SettingsSubsection: React.FC<SettingsSubsectionProps> = ({
    heading,
    description,
    children,
    stretchContent,
    ...rest
}) => (
    <div {...rest} className="mx_SettingsSubsection">
        {typeof heading === "string" ? <SettingsSubsectionHeading heading={heading} /> : <>{heading}</>}
        {!!description && (
            <div className="mx_SettingsSubsection_description">
                <SettingsSubsectionText>{description}</SettingsSubsectionText>
            </div>
        )}
        {!!children && (
            <div
                className={classNames("mx_SettingsSubsection_content", {
                    mx_SettingsSubsection_contentStretch: !!stretchContent,
                    mx_SettingsSubsection_noHeading: !heading && !description,
                })}
            >
                {children}
            </div>
        )}
    </div>
);

export default SettingsSubsection;
