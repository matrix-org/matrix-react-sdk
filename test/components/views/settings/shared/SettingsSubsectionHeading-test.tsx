/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { render } from "@testing-library/react";
import React from "react";

import { SettingsSubsectionHeading } from "../../../../../src/components/views/settings/shared/SettingsSubsectionHeading";

describe("<SettingsSubsectionHeading />", () => {
    const defaultProps = {
        heading: "test",
    };
    const getComponent = (props = {}) => render(<SettingsSubsectionHeading {...defaultProps} {...props} />);

    it("renders without children", () => {
        const { container } = getComponent();
        expect({ container }).toMatchSnapshot();
    });

    it("renders with children", () => {
        const children = <a href="/#">test</a>;
        const { container } = getComponent({ children });
        expect({ container }).toMatchSnapshot();
    });
});
