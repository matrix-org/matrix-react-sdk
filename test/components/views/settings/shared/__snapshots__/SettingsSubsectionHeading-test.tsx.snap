// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsSubsectionHeading /> renders with children 1`] = `
{
  "container": <div>
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h4
        class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
      >
        test
      </h4>
      <a
        href="/#"
      >
        test
      </a>
    </div>
  </div>,
}
`;

exports[`<SettingsSubsectionHeading /> renders without children 1`] = `
{
  "container": <div>
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h4
        class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
      >
        test
      </h4>
    </div>
  </div>,
}
`;
